tools/clang/include/clang/SPIRV/Type.h

```diff
@@ -0,0 +1,127 @@
+//===-- Type.h - SPIR-V Type ------------------------------------*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+#ifndef LLVM_CLANG_SPIRV_TYPE_H
+#define LLVM_CLANG_SPIRV_TYPE_H
+
+#include <set>
+#include <unordered_set>
+#include <vector>
+
+#include "clang/SPIRV/Decoration.h"
+#include "clang/SPIRV/spirv.hpp"
+#include "llvm/ADT/Optional.h"
+
+namespace clang {
+namespace spirv {
+
+class SPIRVContext;
+
+/// \brief SPIR-V Type
+///
+/// This class defines a unique SPIR-V Type.
+/// A SPIR-V Type includes its <opcode> defined by the SPIR-V Spec.
+/// It also incldues any arguments (32-bit words) needed to define the
+/// type. It also includes a set of decorations that are applied to that type.
+///
+/// The class includes static getXXX(...) functions for getting pointers of any
+/// needed type. A unique type has a unique pointer (e.g. calling
+/// 'getBoolean' function will always return the same pointer for the given
+/// context)
+class Type {
+public:
+  spv::Op getOpcode() const { return opcode; }
+  const std::vector<uint32_t> &getArgs() const { return args; }
+  bool isBooleanType() const;
+  bool isIntegerType() const;
+  bool isFloatType() const;
+  bool isNumericalType() const;
+  bool isScalarType() const;
+  bool isVectorType() const;
+  bool isMatrixType() const;
+  bool isArrayType() const;
+  bool isStructureType() const;
+  bool isAggregateType() const;
+  bool isCompositeType() const;
+  bool isImageType() const;
+
+  static const Type *getType(SPIRVContext &ctx, spv::Op op,
```

Comment:
we still do since API users may want to create arbitrary Type objects with arbitrary decorations that they want. However, most often users will be using getInt, getBool, getFloat, etc.